src/hotspot/share/runtime/runtimeUpcalls.hpp

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 1997, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_RUNTIME_UPCALLS_HPP
+#define SHARE_RUNTIME_RUNTIME_UPCALLS_HPP
+
+#include "memory/allStatic.hpp"
+#include "memory/resourceArea.hpp"
+#include "runtime/methodDetails.hpp"
+#include "utilities/macros.hpp"
+
+enum RuntimeUpcallType{
+  onMethodEntry = 0,
+  onMethodExit,
+  numTypes
+};
+
+typedef void (*RuntimeUpcall)(JavaThread* current);
```

Comment:
Not needed for the AOTEndTrainingOnMethodEntry option, but I think passing the method as the argument for which upcall is being done could be beneficial.

Reply:
I can see where that might be useful, but it's complicated as we have a single call back but the 'method' comes back from the interpreter as a ciMethod instance and from c1/c2 as a Method instance, so we'd need to either have a level of indirection so we could wrap the method (like I did with MethodDetails) or have multiple callbacks